### PR TITLE
Allow desktop software to inhibit devices to prevent updates

### PR DESCRIPTION
**Pull request by hughsie**
On edge hardware a process may want to disable firmware updates as it might be a bad time to allow an upgrade.

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

**Review by superm1 — COMMENTED** [review]
Why inhibit individual devices?  Shouldn't the inhibit be for all?

**Comment by hughsie** [comment]
In this case it was intermittently polling a device, don't suggest firmware update -- but another usecase was "using webcam, don't update webcam" -- I don't think we'd want to inhibit all devices for this -- it would also be to easy to make everything "stop getting security updates" from an app point of view.

File: libfwupd/fwupd-client-sync.c [hunk 1]
```diff
@@ -673,6 +673,120 @@ fwupd_client_unlock(FwupdClient *self,
...
+ * by other applications or by the system (e.g. low power state).
+ *
+ * The same application can inhibit a device multiple times.
+ *
+ * Returns: (transfer full): a string, or %NULL for failure
+ *
+ * Since: 1.8.10
+ **/
+gchar *
+fwupd_client_inhibit(FwupdClient *self,
```
Comment (hughsie) [inline libfwupd/fwupd-client-sync.c:703]: @superm1 maybe `fwupd_client_device_inhibit()` would be clearer?

**Comment by superm1** [comment]
> In this case it was intermittently polling a device, don't suggest firmware update -- but another usecase was "using webcam, don't update webcam" -- I don't think we'd want to inhibit all devices for this -- it would also be to easy to make everything "stop getting security updates" from an app point of view.

I feel like it's a lot to expect software to know what devices to inhibit.

I guess walk me through the webcam case. What software would identify it's in use? How would it link it to the fwupd device?

**Comment by hughsie** [comment]
> I guess walk me through the webcam case. What software would identify it's in use? 

The video-conferencing app running in usespace.

> How would it link it to the fwupd device?

 I guessed it would use the InstanceID/GUID, i.e. like `USB\VID_0BDA&PID_5850`

**Comment by superm1** [comment]
> > I guess walk me through the webcam case. What software would identify it's in use? 
> 
> The video-conferencing app running in usespace.

I'm still not convinced. Userspace video conferencing apps will use stuff like /dev/video0 and it means they need to link that to a fwupd device.

> 
> > How would it link it to the fwupd device?
> 
>  I guessed it would use the InstanceID/GUID, i.e. like `USB\VID_0BDA&PID_5850`

So the other worry I have is you don't know the topology of devices. What if my webcam is connected to a USB hub connected to my USB4 dock and I decide to update my host controller's retimer.

The retimer update takes out the whole chain but the software can't possibly know that - we only inhibited the webcam.

We don't show a relationship for retimers today iirc.

**Comment by hughsie** [comment]
Fair point. Do you think it makes sense to support both per-device and whole-device inhibits, or just the latter?

**Comment by superm1** [comment]
> Fair point. Do you think it makes sense to support both per-device and whole-device inhibits, or just the latter?

I was thinking whole device only. 
The other thing to think about is that we have a usable while updating flag. Clients can be smarter on what to update - ie only apply updates for devices without they flag

**Comment by hughsie** [comment]
How about this version?

File: libfwupd/fwupd-client-sync.c [hunk 2]
```diff
@@ -673,6 +673,112 @@ fwupd_client_unlock(FwupdClient *self,
...
+ * @reason: (not nullable): the inhibit reason, e.g. `user active`
+ * @cancellable: (nullable): optional #GCancellable
+ * @error: (nullable): optional return location for an error
+ *
+ * Marks all devices as unavailable for update. Update is only available if there is no other
+ * inhibit imposed by other applications or by the system (e.g. low power state).
+ *
+ * The same application can inhibit a device multiple times.
+ *
+ * Returns: (transfer full): a string, or %NULL for failure
```
Comment (superm1) [inline libfwupd/fwupd-client-sync.c:697]: So this presumably returns a string that you use to undo inhibit, right?

You'll probably want to document it.

File: libfwupd/fwupd-client.c [hunk 3]
```diff
@@ -5390,6 +5390,167 @@ fwupd_client_upload_bytes_finish(FwupdClient *self, GAsyncResult *res, GError **
...
+ * @self: a #FwupdClient
+ * @reason: (not nullable): the inhibit reason, e.g. `user active`
+ * @cancellable: (nullable): optional #GCancellable
+ * @callback: the function to run on completion
+ * @callback_data: the data to pass to @callback
+ *
+ * Marks all devices as unavailable for update. Update is only available if there is no other
+ * inhibit imposed by other applications or by the system (e.g. low power state).
+ *
+ * The same application can inhibit a device multiple times.
```
Comment (superm1) [inline libfwupd/fwupd-client.c:5424]: Not applicable

File: src/fu-util-common.c [hunk 4]
```diff
@@ -1402,6 +1402,10 @@ fu_util_device_problem_to_string(FwupdClient *client, FwupdDevice *dev, FwupdDev
 		/* TRANSLATORS: The device cannot be updated due to missing vendor's license." */
 		return g_strdup(_("Device requires a software license to update"));
 	}
+	if (problem == FWUPD_DEVICE_PROBLEM_SYSTEM_INHIBIT) {
+		/* TRANSLATORS: an application is preventing system updates */
+		return g_strdup(_("System is inhibited"));
```
Comment (superm1) [inline src/fu-util-common.c:1407]: I think this string should be expanded.

File: src/fu-util.c [hunk 5]
```diff
@@ -977,6 +977,23 @@ fu_util_device_test_filename(FuUtilPrivate *priv,
...
+static gboolean
+fu_util_inhibit(FuUtilPrivate *priv, gchar **values, GError **error)
+{
+	const gchar *reason = "not set";
+	g_autofree gchar *inhibit_id = NULL;
+
+	if (g_strv_length(values) > 0)
+		reason = values[1];
+
+	/* inhibit, wait, uninhibit */
```
Comment (superm1) [inline src/fu-util.c:989]: This doesn't make sense to me. It should be two separate commands, or if combined command needs to have the delay as an argument.

**Review by superm1 — COMMENTED** [review]
I think generally this direction makes sense. But if it's going to be enabled by command line there should probably be two separate commands or at least an argument for how long to inhibit.

File: data/bash-completion/fwupdmgr [hunk 6]
```diff
@@ -20,6 +20,7 @@ _fwupdmgr_cmd_list=(
 	'get-updates'
 	'get-upgrades'
 	'get-plugins'
+	'inhibit'
```
[the remark this answers is not in the document]
Comment (hughsie) [reply data/bash-completion/fwupdmgr:23]: Done.

File: libfwupd/fwupd-client-sync.c [hunk 7]
```diff
@@ -673,6 +673,112 @@ fwupd_client_unlock(FwupdClient *self,
...
+ * fwupd_client_inhibit:
+ * @self: a #FwupdClient
+ * @reason: (not nullable): the inhibit reason, e.g. `user active`
+ * @cancellable: (nullable): optional #GCancellable
+ * @error: (nullable): optional return location for an error
+ *
+ * Marks all devices as unavailable for update. Update is only available if there is no other
+ * inhibit imposed by other applications or by the system (e.g. low power state).
+ *
+ * The same application can inhibit a device multiple times.
```
[the remark this answers is not in the document]
Comment (hughsie) [reply libfwupd/fwupd-client-sync.c:695]: Fixed.

File: libfwupd/fwupd-client-sync.c (hunk 2 above)
Comment (hughsie) [reply libfwupd/fwupd-client-sync.c:697]: Fixed.

File: libfwupd/fwupd-client.c (hunk 3 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:5424]: Fixed.

File: src/fu-util-common.c (hunk 4 above)
Comment (hughsie) [reply src/fu-util-common.c:1407]: Got any ideas? `All devices are prevented from update by inhibit` perhaps?

File: src/fu-util.c (hunk 5 above)
Comment (hughsie) [reply src/fu-util.c:989]: Agreed, two separate commands makes a lot more sense if we put in place a cancellable GMainLoop. Fixed.

File: src/fu-util-common.c (hunk 4 above)
Comment (superm1) [reply src/fu-util-common.c:1407]: sounds good

File: src/fu-util.c [hunk 8]
```diff
@@ -977,6 +978,40 @@ fu_util_device_test_filename(FuUtilPrivate *priv,
...
+	g_autofree gchar *inhibit_id = NULL;
+
+	if (g_strv_length(values) > 0)
+		reason = values[1];
+
+	/* inhibit then wait */
+	inhibit_id = fwupd_client_inhibit(priv->client, reason, priv->cancellable, error);
+	if (inhibit_id == NULL)
+		return FALSE;
+	g_print("Inhibit ID is: %s, use ctrl-c to abort\n", inhibit_id);
```
Comment (superm1) [inline src/fu-util.c:994]: presumably this should be translatable

Comment (hughsie) [reply src/fu-util.c:994]: How about this:

Screenshot from 2023-01-23 09-32-17

Comment (superm1) [reply src/fu-util.c]: Lgtm

**Pull request merged**